src/raphson.jl

```diff
@@ -172,6 +182,43 @@ function perform_step!(cache::NewtonRaphsonCache{false})
     return nothing
 end
 
+function objective_linesearch(cache::NewtonRaphsonCache) ## returns the objective functions required in LS
+    @unpack f, p = cache
+
+    function fo(x)
+        return dot(value_f(cache, x), value_f(cache, x)) / 2
+    end
+
+    function g!(g_o, x)
+        mul!(g_o, simple_jacobian(cache, x)', value_f(cache, x)) 
+        g_o
+    end
+
+    function fg!(g_o, x)
+        g!(g_o, x)
+        fo(x)
+    end
+    return fo, g!, fg!
+end
+
+function perform_linesearch!(cache::NewtonRaphsonCache)
+    @unpack u, fu, du1, alg, α, g_o = cache
+    fo, g!, fg! = objective_linesearch(cache)
+    ϕ(α) = fo(u .- α .* du1)
+
+    function dϕ(α)
+        g!(g_o, u .- α .* du1)
+        return dot(g_o, -du1)
+    end
+
+    function ϕdϕ(α)
+        return (fg!(g_o, u .- α .* du1), dot(g_o, -du1))
+    end
```

Comment:
Similarly, these 3 functions are defined each time the `perform_linesearch!` function is called. Can this lead to performance issues?